src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -401,6 +501,162 @@ private static void AssertNoImplicitInterpolatedStringHandlerConversions(Immutab
+        {
+            var result = ArrayBuilder<BoundExpression>.GetInstance();
+            var currentStringConst = PooledStringBuilder.GetInstance();
+            (BoundCall? AppendLiteralCall, BoundLiteral? StringLiteral) reusable = default;
+            BoundExpression? lastAppendReceiver = null;
+            int filledHolesCount = 0;
+            int increasedLiteralLength = 0;
+
+            BoundExpression createAppendLiteralCallFromBoundLiteral(BoundLiteral literal, BoundExpression receiver) =>
+                _factory.InstanceCall(receiver, BoundInterpolatedString.AppendLiteralMethod, literal);
+            BoundExpression createAppendLiteralCallFromString(string literal, BoundExpression receiver) =>
+                createAppendLiteralCallFromBoundLiteral(_factory.StringLiteral(literal), receiver);
+            void mergeReusableIntoBuilder()
+            {
+                if (reusable.StringLiteral.ConstantValueOpt != null)
+                {
+                    currentStringConst.Builder.Append(reusable.StringLiteral.ConstantValueOpt.StringValue);
+                }
+                reusable = default;
+            }
+            void mergeReusableIntoResult(BoundCall? call = null)
+            {
+                switch (reusable)
+                {
+                    case (not null, _):
+                        result.Add(reusable.AppendLiteralCall);
+                        break;
+                    case (_, not null):
+                        Debug.Assert(call?.ReceiverOpt is not null || lastAppendReceiver is not null);
+                        result.Add(createAppendLiteralCallFromBoundLiteral(reusable.StringLiteral, call?.ReceiverOpt ?? lastAppendReceiver!));
+                        break;
+                    default:
+                        if (currentStringConst.Length > 0)
+                        {
+                            Debug.Assert(call?.ReceiverOpt is not null || lastAppendReceiver is not null);
+                            result.Add(createAppendLiteralCallFromString(currentStringConst.ToStringAndFree(), call?.ReceiverOpt ?? lastAppendReceiver!));
+                            currentStringConst = PooledStringBuilder.GetInstance();
+                        }
+                        return;
+                }
+                reusable = default;
+            }
+
+            foreach (var part in parts)
+            {
+                if (part is not BoundCall { Method.Name: not null, } call)
+                {
+                    mergeReusableIntoResult();
+                    result.Add(part);
+                    continue;
+                }
+
+                switch (call.Method.Name)
+                {
+                    case BoundInterpolatedString.AppendFormattedMethod:
+                        {
+                            // never be BoundLiteral (can be BoundLocal), so see ConstantValueOpt instead
```

Comment:
What will never be a `BoundLiteral`? The argument certainly can be, and that's the only thing I can think of that you might be saying? Are you saying it will never be a `BoundLiteral` of a string?

Reply:
> Are you saying it will never be a BoundLiteral of a string?

I forgot whether or not, but I think I meant so at that time.
I will fix the comment.